pennylane/transforms/decompositions/clifford_t/clifford_t_transform.py

```diff
@@ -0,0 +1,479 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Transform function for the Clifford+T decomposition."""
+
+import math
+import warnings
+from itertools import product
+from typing import Sequence, Callable
+
+import pennylane as qml
+from pennylane.queuing import QueuingManager
+from pennylane.transforms.core import transform
+from pennylane.tape import QuantumTape
+from pennylane.transforms.optimization import (
+    cancel_inverses,
+    commute_controlled,
+    merge_rotations,
+    remove_barrier,
+)
+from pennylane.transforms.optimization.optimization_utils import find_next_gate, _fuse_global_phases
+from .solovay_kitaev import sk_decomposition
+
+# Single qubits Clifford+T gates in PL
+_CLIFFORD_T_ONE_GATES = [
+    qml.Identity,
+    qml.PauliX,
+    qml.PauliY,
+    qml.PauliZ,
+    qml.Hadamard,
+    qml.S,
+    qml.SX,
+    qml.T,
+]
+
+# Two qubits Clifford+T gates in PL
+_CLIFFORD_T_TWO_GATES = [
+    qml.CNOT,
+    qml.CY,
+    qml.CZ,
+    qml.SWAP,
+    qml.ISWAP,
+]
+
+# Single-parameter gates in PL
+_PARAMETER_GATES = (qml.RX, qml.RY, qml.RZ, qml.Rot, qml.PhaseShift)
+
+# Clifford+T gate set
+_CLIFFORD_T_GATES = tuple(_CLIFFORD_T_ONE_GATES + _CLIFFORD_T_TWO_GATES) + (qml.GlobalPhase,)
+
+
+def check_clifford_op(op):
+    r"""Checks if an operator is Clifford or not.
+
+    For a given unitary operator :math:`U` acting on :math:`N` qubits, this method checks that the
+    transformation :math:`UPU^{\dagger}` maps the Pauli tensor products :math:`P = {I, X, Y, Z}^{\otimes N}`
+    to Pauli tensor products with :math:`O(N \times 8^N)` time complexity when using naive matrix multiplication.
+
+    Args:
+        op: the operator that needs to be tested
+
+    Returns:
+        Bool that represents whether the provided operator is Clifford or not.
+    """
+
+    # Check if matrix can be calculated for the operator
+    if not op.has_matrix:
+        return False
```

Comment:
So here do we risk the possibility of false negatives? I.e., ops that do not have a matrix method but are Clifford?

Reply:
If an operator doesn't have a matrix method, we won't be able to check UPU^\dagger for them. An alternative is to check for their decomposition, but if an operator can have a known decomposition method, I guess it should also have a matrix one.

Reply:
to chime in - we only reach this point if the op is not an instance of any pre-defined clifford+T gate, or parametrized gate with a param close to a non-parametrized clifford op (eg. RZ(math.pi))

Reply:
Resolution - this function is specific to this pipeline and not a general checker of Cliffordness